File: server/src/models/VerseActivity.js
const mongoose = require('mongoose');

const verseActivitySchema = new mongoose.Schema({
  chapterId: {
    type: String,
    required: true,
    index: true,
  },
  verseNumber: {
    type: Number,
    required: true,
    index: true,
  },
  title: {
    type: String,
    required: true,
  },
  description: {
    type: String,
    required: true,
  },
  type: {
    type: String,
    enum: ['experiment', 'meditation', 'reflection', 'practice', 'journaling', 'daily_challenge'],
    required: true,
  },
  instructions: [{
    step: {
      type: Number,
      required: true,
    },
    description: {
      type: String,
      required: true,
    },
  }],
  duration: {
    type: String, // e.g., "5 minutes", "1 day", "1 week"
    default: '',
  },
  difficulty: {
    type: String,
    enum: ['beginner', 'intermediate', 'advanced'],
    default: 'beginner',
  },
  benefits: [{
    type: String,
  }],
  iconUrl: {
    type: String,
    default: '',
  },
  order: {
    type: Number,
    default: 0,
  },
}, {
  timestamps: true,
  toJSON: { 
    virtuals: true,
    transform: function(doc, ret) {
      ret.id = `${ret.chapterId}-${ret.verseNumber}-${ret.order}`;
      return ret;
    }
  },
  toObject: { virtuals: true },
});

// Compound index for efficient queries
verseActivitySchema.index({ chapterId: 1, verseNumber: 1, order: 1 });

module.exports = mongoose.model('VerseActivity', verseActivitySchema);

File: geeta book app/lib/shared/domain/models/verse_activity.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'verse_activity.freezed.dart';
part 'verse_activity.g.dart';

enum VerseActivityType {
  experiment,
  meditation,
  reflection,
  practice,
  journaling,
  @JsonValue('daily_challenge')
  dailyChallenge,
}

enum ActivityDifficulty {
  beginner,
  intermediate,
  advanced,
}

@freezed
class ActivityInstruction with _$ActivityInstruction {
  const factory ActivityInstruction({
    required int step,
    required String description,
  }) = _ActivityInstruction;

  factory ActivityInstruction.fromJson(Map<String, dynamic> json) =>
      _$ActivityInstructionFromJson(json);
}

@freezed
class VerseActivity with _$VerseActivity {
  const factory VerseActivity({
    required String id,
    required String chapterId,
    required int verseNumber,
    required String title,
    required String description,
    required VerseActivityType type,
    required List<ActivityInstruction> instructions,
    required String duration,
    required ActivityDifficulty difficulty,
    required List<String> benefits,
    String? iconUrl,
    required int order,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) = _VerseActivity;

  factory VerseActivity.fromJson(Map<String, dynamic> json) =>
      _$VerseActivityFromJson(json);
}
